src/engine/fileOps/tfm.cpp

```diff
@@ -0,0 +1,900 @@
+      if (maxPat<orderList[i]) maxPat=orderList[i];
+
+      for (int j=0; j<6; j++) {
+        ds.subsong[0]->orders.ord[j][i]=orderList[i];
+        ds.subsong[0]->pat[j].data[orderList[i]]=new DivPattern;
+      }
+    }
+
+    DivInstrument* insMaps[256];
+    int insNumMaps[256];
+
+    // instrument names
+    logD("parsing instruments");
+    unsigned char insName[16];
+    int insCount=0;
+    for (int i=0; i<255; i++) {
+      reader.read(insName,16);
+
+      if (memcmp(insName,"\xFF\xFF\xFF\xFF\xFF\xFF\xFF\xFF\xFF\xFF\xFF\xFF\xFF\xFF\xFF\xFF",16)==0) {
+        logD("instrument unused");
+        insNumMaps[i]=i;
+        insMaps[i]=NULL;
+        continue;
+      }
+
+      DivInstrument* ins=new DivInstrument;
+      ins->type=DIV_INS_FM;
+      ins->name=String((const char*)insName,strnlen((const char*)insName,16));
+      ds.ins.push_back(ins);
+
+      insNumMaps[i]=insCount;
+      insCount++;
+
+      insMaps[i]=ins;
+    }
+
+    ds.insLen=insCount;
+
+    // instrument data
+    for (int i=0; i<255; i++) {
+      if (!insMaps[i]) {
+        reader.skip(42);
+        continue;
+      }
+
+      insMaps[i]->fm.alg=reader.readC();
+      insMaps[i]->fm.fb=reader.readC();
+
+      for (int j=0; j<4; j++) {
+        insMaps[i]->fm.op[j].mult=reader.readC();
+        insMaps[i]->fm.op[j].dt=reader.readC();
+        insMaps[i]->fm.op[j].tl=reader.readC()^0x7F;
+        insMaps[i]->fm.op[j].rs=reader.readC();
+        insMaps[i]->fm.op[j].ar=reader.readC();
+        insMaps[i]->fm.op[j].dr=reader.readC();
+        insMaps[i]->fm.op[j].d2r=reader.readC();
+        insMaps[i]->fm.op[j].rr=reader.readC();
+        insMaps[i]->fm.op[j].sl=reader.readC();
+        insMaps[i]->fm.op[j].ssgEnv=reader.readC();
+      }
+    }
+
+    ds.notes=notes;
+
+    unsigned char patLens[256];
+    int maxPatLen=0;
+    reader.read(patLens, 256);
+    for (int i=0; i<256; i++) {
+      if (patLens[i]==0) {
+        maxPatLen=256;
+        break;
+      } else if (patLens[i]>maxPatLen) {
+        maxPatLen=patLens[i];
+      }
+    }
+
+    ds.subsong[0]->patLen=maxPatLen;
+
+    struct TFMParsePatternInfo info;
+    info.ds=&ds;
+    info.insNumMaps=insNumMaps;
+    info.maxPat=maxPat;
+    info.patExists=patExists;
+    info.orderList=orderList;
+    info.speedEven=speed>>4;
+    info.speedOdd=speed&0xF;
+    info.interleaveFactor=interleaveFactor;
+    info.patLens=patLens;
+    info.reader=&reader;
+    info.v2=false;
+    TFMParsePattern(info);
+
+    if (active) quitDispatch();
+    BUSY_BEGIN_SOFT;
+    saveLock.lock();
+    song.unload();
+    song=ds;
+    changeSong(0);
+    recalcChans();
+    saveLock.unlock();
+    BUSY_END;
+    if (active) {
+      initDispatch();
+      BUSY_BEGIN;
+      renderSamples();
+      reset();
+      BUSY_END;
+    }
+    success=true;
+  } catch(TFMEndOfFileException& e) {
+    lastError="incomplete file!";
+  } catch(InvalidHeaderException& e) {
+    lastError="invalid info header!";
+  }
+
```

Comment:
Nooooooooo..... Memory leak...

Reply:
What about it that leaks memory? The `file` gets deallocated regardless in the lines afterwards, before `return`

Reply:
It doesn't. It only does in TFMv2. TFMv1 does not have the deallocation.

Reply:
Oh whoops